# Phase 2 Code Challenge

This code challenge is designed to test your understanding of the Phase 2 material. It covers:

- Normal Distribution
- Statistical Tests
- Bayesian Statistics
- Linear Regression

_Read the instructions carefully_. You will be asked both to write code and to answer short answer questions.

## Code Tests

We have provided some code tests for you to run to check that your work meets the item specifications. Passing these tests does not necessarily mean that you have gotten the item correct - there are additional hidden tests. However, if any of the tests do not pass, this tells you that your code is incorrect and needs changes to meet the specification. To determine what the issue is, read the comments in the code test cells, the error message you receive, and the item instructions.

## Short Answer Questions 

For the short answer questions...

* _Use your own words_. It is OK to refer to outside resources when crafting your response, but _do not copy text from another source_.

* _Communicate clearly_. We are not grading your writing skills, but you can only receive full credit if your teacher is able to fully understand your response. 

* _Be concise_. You should be able to answer most short answer questions in a sentence or two. Writing unnecessarily long answers increases the risk of you being unclear or saying something incorrect.

In [1]:
# Run this cell without changes to import the necessary libraries

import itertools
import numpy as np
import pandas as pd 
from numbers import Number
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import pickle

---
## Part 1: Normal Distribution [Suggested time: 20 minutes]
---
In this part, you will analyze check totals at a TexMex restaurant. We know that the population distribution of check totals for the TexMex restaurant is normally distributed with a mean of \\$20 and a standard deviation of \\$3. 

### 1.1) Create a numeric variable `z_score_26` containing the z-score for a \\$26 check. 

**Starter Code**

    z_score_26 = 

In [24]:
# your code here
from scipy.stats import norm
mean = 20
s = 3
z_score_26 = (26 - 20) / 3
z_score_26

2.0

In [25]:
# This test confirms that you have created a numeric variable named z_score_26

assert isinstance(z_score_26, Number)


### 1.2) Create a numeric variable `p_under_26` containing the approximate proportion of all checks that are less than \\$26.

Hint: Use the answer from the previous question along with the empirical rule, a Python function, or this [z-table](https://www.math.arizona.edu/~rsims/ma464/standardnormaltable.pdf).

**Starter Code**

    p_under_26 = 

In [28]:
# your code here
p_under_26 = norm.cdf(26, loc = 20, scale = 3)
p_under_26

0.9772498680518208

In [27]:
# This test confirms that you have created a numeric variable named p_under_26

assert isinstance(p_under_26, Number)

# These tests confirm that p_under_26 is a value between 0 and 1

assert p_under_26 >= 0
assert p_under_26 <= 1


### 1.3) Create numeric variables `conf_low` and `conf_high` containing the lower and upper bounds (respectively) of a 95% confidence interval for the mean of one waiter's check amounts using the information below. 

One week, a waiter gets 100 checks with a mean of \\$19 and a standard deviation of \\$3.

**Starter Code**

    n = 100
    mean = 19
    std = 3
    
    conf_low = 
    conf_high = 

In [31]:
# your code here
mean = 19
sd = 3
z = 1.96
n = 100
meo = z * sd / np.sqrt(n)
conf_low, conf_high = (mean-meo , mean+meo)
print(conf_low, conf_high)

18.412 19.588


In [32]:
# These tests confirm that you have created numeric variables named conf_low and conf_high

assert isinstance(conf_low, Number)
assert isinstance(conf_high, Number)

# This test confirms that conf_low is below conf_high

assert conf_low < conf_high

# These statements print your answers for reference to help answer the next question

print('The lower bound of the 95% confidence interval is {}'.format(conf_low))
print('The upper bound of the 95% confidence interval is {}'.format(conf_high))


The lower bound of the 95% confidence interval is 18.412
The upper bound of the 95% confidence interval is 19.588


### 1.4) Short Answer: Interpret the 95% confidence interval you just calculated in Question 1.3.

18.412 to 19.588

---
## Part 2: Statistical Testing [Suggested time: 20 minutes]
---
The TexMex restaurant recently introduced queso to its menu.

We have a random sample containing 2000 check totals, all from different customers: 1000 check totals for orders without queso ("no queso") and 1000 check totals for orders with queso ("queso").

In the cell below, we load the sample data for you into the arrays `no_queso` and `queso` for the "no queso" and "queso" order check totals, respectively.

In [33]:
# Run this cell without changes

# Load the sample data 
no_queso = pickle.load(open('./no_queso.pkl', 'rb'))
queso = pickle.load(open('./queso.pkl', 'rb'))

### 2.1) Short Answer: State null and alternative hypotheses to use for testing whether customers who order queso spend different amounts of money from customers who do not order queso.

In [35]:
print(no_queso.mean(), queso.mean())

20.047532509358607 23.992814607054243


H0: customers who order queso spend less or equal customers who do not order queso
H1: customers who order queso spend more than customers who do not order queso

### 2.2) Short Answer: What would it mean to make a Type I error for this specific hypothesis test?

Your answer should be _specific to this context,_  not a general statement of what Type I error is.

z-test. we expect customers who order queso spend less than customers who do not order queso, Actually customers who order queso spend more than customers who do not order queso

### 2.3) Create a numeric variable `p_value` containing the p-value associated with a statistical test of your hypotheses. 

You must identify and implement the correct statistical test for this scenario. You can assume the two samples have equal variances.

Hint: Use `scipy.stats` to calculate the answer - it has already been imported as `stats`. Relevant documentation can be found [here](https://docs.scipy.org/doc/scipy/reference/stats.html#statistical-tests).

**Starter Code**

    p_value = 

In [67]:
# your code here
no_queso_mean = np.mean(no_queso)
queso_mean = np.mean(queso)
sigma = np.std(no_queso)
n=1000
order_z = (queso_mean - no_queso_mean) / (sigma / np.sqrt(n))
p_value = 1-stats.norm.cdf(order_z)

In [79]:
p_value = ttest_ind(queso, no_queso)[1]

In [80]:
p_value

1.29670967092511e-307

In [74]:
# These tests confirm that you have created a numeric variable named p_value

assert isinstance(p_value, Number)


### 2.4) Short Answer: Can you reject the null hypothesis using a significance level of $\alpha$ = 0.05? Explain why or why not.

reject null, because p less than alpha. Meaning it is big different between no_queso and queso. H1: customers who order queso spend more than customers who do not order queso

---
## Part 3: Bayesian Statistics [Suggested time: 15 minutes]
---
A medical test is designed to diagnose a certain disease. The test has a false positive rate of 10%, meaning that 10% of people without the disease will get a positive test result. The test has a false negative rate of 2%, meaning that 2% of people with the disease will get a negative result. Only 1% of the population has this disease.

### 3.1) Create a numeric variable `p_pos_test` containing the probability of a person receiving a positive test result.

Assume that the person being tested is randomly selected from the broader population.

**Starter Code**
    
    false_pos_rate = 0.1
    false_neg_rate = 0.02
    population_rate = 0.01
    
    p_pos_test = 

In [54]:
# your code here
false_pos_rate = 0.1
false_neg_rate = 0.02
population_rate = 0.01
    
p_pos_test = population_rate * (1-false_neg_rate) + (1 - population_rate) * false_pos_rate
p_pos_test

0.10880000000000001

In [55]:
# This test confirms that you have created a numeric variable named p_pos_test

assert isinstance(p_pos_test, Number)

# These tests confirm that p_pos_test is a value between 0 and 1

assert p_pos_test >= 0
assert p_pos_test <= 1


### 3.2) Create a numeric variable `p_pos_pos` containing the probability of a person actually having the disease if they receive a positive test result.

Assume that the person being tested is randomly selected from the broader population.

Hint: Use your answer to the previous question to help answer this one.

**Starter Code**
    
    false_pos_rate = 0.1
    false_neg_rate = 0.02
    population_rate = 0.01
    
    p_pos_pos = 

In [61]:
# your code here
p_pos_pos = 0.98 * population_rate / p_pos_test

In [62]:
p_pos_pos

0.09007352941176469

In [63]:
# This test confirms that you have created a numeric variable named p_pos_pos

assert isinstance(p_pos_pos, Number)

# These tests confirm that p_pos_pos is a value between 0 and 1

assert p_pos_pos >= 0
assert p_pos_pos <= 1


---

## Part 4: Linear Regression [Suggested Time: 20 min]
---

In this section, you'll run regression models with advertising data. The dataset includes the advertising spending and sales for 200 products. These products are a random sample of 200 products from a larger population of products. 

The dataset has three features - `TV`, `radio`, and `newspaper` - that describe how many thousands of advertising dollars were spent promoting the product. The target, `sales`, describes how many millions of dollars in sales the product had.

In [2]:
# Run this cell without changes

import statsmodels
from statsmodels.formula.api import ols

data = pd.read_csv('./advertising.csv').drop('Unnamed: 0', axis=1)
X = data.drop('sales', axis=1)
y = data['sales']

### 4.1) Create a variable `tv_mod_summ` containing the summary of a fit StatsModels `ols` model for a linear regression using `TV` to predict `sales`. 

**Starter Code**

    tv_mod = ols(
    tv_mod_summ = tv_mod.summary()

In [4]:
data.describe()

TV       radio   newspaper       sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   14.022500
std     85.854236   14.846809   21.778621    5.217457
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   10.375000
50%    149.750000   22.900000   25.750000   12.900000
75%    218.825000   36.525000   45.100000   17.400000
max    296.400000   49.600000  114.000000   27.000000

In [9]:
# your code here
import statsmodels.api as sm
y = data['sales']
x = data['TV']
tv_mod = sm.OLS(y, x).fit()
tv_mod_summ = tv_mod.summary()

In [10]:
# This test confirms that you have created a variable named tv_mod_summ containing a StatsModels Summary

assert type(tv_mod_summ) == statsmodels.iolib.summary.Summary

# This statement prints your answer for reference to help answer the next question

print(tv_mod_summ)


                                 OLS Regression Results                                
Dep. Variable:                  sales   R-squared (uncentered):                   0.897
Model:                            OLS   Adj. R-squared (uncentered):              0.896
Method:                 Least Squares   F-statistic:                              1733.
Date:                Thu, 24 Mar 2022   Prob (F-statistic):                   3.52e-100
Time:                        17:54:15   Log-Likelihood:                         -597.51
No. Observations:                 200   AIC:                                      1197.
Df Residuals:                     199   BIC:                                      1200.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

### 4.2) Short Answer: Is there a statistically significant relationship between TV advertising spend and sales in this model? How did you determine this from the model output? 

This question is asking you to use your findings from the sample in your dataset to make an inference about the relationship between TV advertising spend and sales in the broader population.

R-squared is 0.897, it has significant relationship between TV advertising spend and sales in this model

### 4.3) Short Answer: Run the cell below to produce a correlation matrix. Given the output, would you expect any collinearity issues if you included all of these features in one regression model? 

In [11]:
# Run this cell

X.corr()

TV     radio  newspaper
TV         1.000000  0.054809   0.056648
radio      0.054809  1.000000   0.354104
newspaper  0.056648  0.354104   1.000000

It is no collinearity issues, because all correlation lower than 0.06

### 4.4) Create a variable `all_mod_summ` containing the summary of a fit StatsModels `ols` model for a multiple regression using `TV`, `radio`, and `newspaper` to predict `sales`. 

**Starter Code**

    all_mod = ols(
    all_mod_summ = all_mod.summary()

In [17]:
# your code here
y = data['sales']
multiple_formula = sm.add_constant(data[['TV', 'radio', 'newspaper']])
all_mod = sm.OLS(y, multiple_formula).fit()
all_mod_summ = all_mod.summary()

In [18]:
# This test confirms that you have created a variable named all_mod_summ containing a StatsModels Summary

assert type(all_mod_summ) == statsmodels.iolib.summary.Summary

# This statement prints your answer for reference to help answer the next question

print(all_mod_summ)


                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Thu, 24 Mar 2022   Prob (F-statistic):           1.58e-96
Time:                        18:03:03   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.0

### 4.5) Short Answer: Does this model do a better job of explaining sales than the previous model using only the `TV` feature? Explain how you determined this based on the model output. 

Yes, TV has the hightest number in t 32.809, radio second.